README.md
# PVOutput

[![Actions Status](https://github.com/johnf/pvoutput/workflows/Test/badge.svg)](https://github.com/johnf/pvoutput/actions)
[![Coverage Status](https://coveralls.io/repos/johnf/pvoutput/badge.svg?branch=master&service=github)](https://coveralls.io/github/johnf/pvoutput?branch=master)
[![Gem Version](https://badge.fury.io/rb/pvoutput.svg)](http://badge.fury.io/rb/pvoutput)


Ruby library for talking to the PVOutput API.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'pvoutput'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install pvoutput

## Usage

In order to use pvoutput in your application you need to use

```ruby
require 'pvoutput/client'
```

First step is to create a PVOutput client using your PVOutput assigned system_id and api_key

```ruby
pvoutput = PVOutput::Client.new(system_id, api_key)
```

At the moment you [donate](http://pvoutput.org/donate.jsp) to PVOutput you can enable the donation mode by creating a PVOutput client using

```ruby
pvoutput = PVOutput::Client.new(system_id, api_key, true)
```

Now you can report your real time status to PVOutput using

```ruby
pvoutput.add_status(
  options
)
```

The add_status operation accepts the following options

| Option           | PVOutput Parameter |
| ---------------- | ------------------ |
| energy_generated | v1 |
| power_generated  | v2 |
| energy_consumed  | v3 |
| power_consumed   | v4 |
| temperature      | v5 |
| voltage          | v6 |
| cumulative       | c1 |
| net              | n |

As example

```ruby
client.add_status(
  :energy_generated => 100,
  :power_generated  => 50,
  :temperature      => 30,
  :voltage          => 200,
)
```

You can report your daily output to PVOutput using

```ruby
pvoutput.add_output(
  options
)
```

The add_output operation accepts the following options

| Option                | PVOutput Parameter |
| --------------------- | ------------------ |
| output_date           | d |
| energy_generated      | g |
| peak_power            | pp |
| peak_time             | pt |
| condition             | cd |
| min_temp              | tm |
| max_temp              | tx |
| comments              | cm |
| import_peak           | ip |
| import_off_peak       | io |
| import_shoulder       | is |
| import_high_shoulder  | ih |
| consumption           | c |

As example

```ruby
client.add_output(
  :output_date      => '20160228'
  :energy_generated  => 15000,
  :max_temp      => 30
)
```

You can report also a batch of daily output values to PVOutput using

```ruby
pvoutput.add_batch_output(
  options
)
```

The add_batch_output operation accepts a hash with the date as key and within that the following options

| Option                | PVOutput Parameter |
| --------------------- | ------------------ |
| energy_generated      | g |
| peak_power            | pp |
| peak_time             | pt |
| condition             | cd |
| min_temp              | tm |
| max_temp              | tx |
| comments              | cm |
| import_peak           | ip |
| import_off_peak       | io |
| import_shoulder       | is |
| import_high_shoulder  | ih |
| consumption           | c |

As example

```ruby
client.add_output(
  :'20150101' => {
    :energy_generated => 1239, },
  :'20150102' => {
    :energy_generated => 1523 },
  :'20150103' => {
    :energy_generated => 2190 },
)
```

## Debugging

You can enable HTTParty debugging to see the requests by setting `PVOUTPUT_DEBUG=true`

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake rspec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`.

## Release

To release a new version:
  * Create a branch for the release
  * Update the version number in `version.rb`
  * Update CHANGELOG.md
  * git commit -pm 'Release 1.x.x'
  * git tag v1.x.x
  * git push
  * git push --tags
  * In GitHub add the Change Log section to the release and save it
  * Check everything looks OK in GitHub Actions

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/johnf/pvoutput. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



CHANGELOG.md
# Changelog

# 1.1.0 (2021-07-12)

* add_batch_output is now using addoutput with a batch size of 1 in regular mode and 100 in donation mode (@jwillemsen)

# Changelog

# 1.0.0 (2021-02-14)

* Require a minimum of Ruby 2.5 (@jwillemsen)
* Clean up ruby gem publish GitHub Action (@jwillemsen)

# 0.6.0 (2021-02-06)

This test has no code changes but tests the new release process.

* Move tests from CircleCI to GitHub Actions (@johnf)
* Keep an eye on dependencies with Dependabot (@jwillemsen)
* Schedule regular tests against multiple rubys (@jwillemsen)

# 0.5.0 (2020-09-20)

* Add more detail to error messages (@jwillemsen)
* Support debugging (@johnf)

# 0.4.0 (20160911)

* Add retries to batch_output (@jwillemsen)

# 0.3.0 (20160229)

* Add donation mode (@jwillemsen)
* Add batch_output support (@jwillemsen)

# 0.2.0 (20160227)

* Fix typo in temperature (@jwillemsen)
* Add usage documentation (@jwillemsen)
* Add the add_ouput method for end of day reporting (@jwillemsen)
* Use latest bundler in CircleCI (@johnf)

# 0.1.1 (20151122)

* Convert the system id to a string (@johnf)

# 0.1.0 (20151122)

* Initial Release (@johnf)


spec/spec_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)

require 'coveralls'

Coveralls.wear!

module Kernel
  alias old_sleep sleep
  def sleep(seconds)
    old_sleep(seconds * 0.01)
  end
end


spec/pvoutput_spec.rb
# frozen_string_literal: true

require 'spec_helper'

# rubocop:disable RSpec/FilePath
describe PVOutput do
  it 'has a version number' do
    expect(PVOutput::VERSION).not_to be nil
  end
end
# rubocop:enable RSpec/FilePath


spec/pvoutput/client_spec.rb
# frozen_string_literal: true

require 'spec_helper'

require 'webmock/rspec'
require 'timecop'

require 'pvoutput/client'

# rubocop:disable Metrics/BlockLength,RSpec/FilePath
describe PVOutput::Client do
  describe 'API' do
    let(:system_id) { '1234' }
    let(:api_key) { 'secret' }
    let(:client) { described_class.new(system_id, api_key) }
    let(:donation_client) { described_class.new(system_id, api_key, true) }
    let(:headers) do
      {
        'X-Pvoutput-Apikey'   => 'secret',
        'X-Pvoutput-Systemid' => '1234',
      }
    end

    before do
      Timecop.freeze(Time.local(2015))
    end

    after do
      Timecop.return
    end

    it 'adds a status' do
      body = 'd=20150101&t=00%3A00&v1=100&v2=50&v5=30&v6=200'
      st = stub_request(:post, 'http://pvoutput.org/service/r2/addstatus.jsp').with(:body => body, :headers => headers)

      client.add_status(
        :energy_generated => 100,
        :power_generated  => 50,
        :temperature      => 30,
        :voltage          => 200
      )

      expect(st).to have_been_requested
    end

    it 'adds an output' do
      body = 'd=2015-01-01&g=100&pp=50&tx=45&cm=moo'
      st = stub_request(:post, 'http://pvoutput.org/service/r2/addoutput.jsp').with(:body => body, :headers => headers)

      client.add_output(
        :output_date      => Date.today,
        :energy_generated => 100,
        :peak_power       => 50,
        :max_temp         => 45,
        :comments         => 'moo'
      )

      expect(st).to have_been_requested
    end

    it 'adds a batch output without donation mode' do
      st1 = stub_request(:post, 'http://pvoutput.org/service/r2/addoutput.jsp').with(
        :body => 'data=20150101%2C1239%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C', :headers => headers
      )
      st2 = stub_request(:post, 'http://pvoutput.org/service/r2/addoutput.jsp').with(
        :body => 'data=20150102%2C1523%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C', :headers => headers
      )

      client.add_batch_output(
        :'20150101' => { :energy_generated => 1239 },
        :'20150102' => { :energy_generated => 1523 }
      )

      expect(st1).to have_been_requested
      expect(st2).to have_been_requested
    end

    it 'adds a batch output output with donation mode' do
      body = 'data=20150101%2C1239%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C%3B20150102%2C1523%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C'
      st = stub_request(:post, 'http://pvoutput.org/service/r2/addoutput.jsp').with(
        :body => body, :headers => headers
      )

      donation_client.add_batch_output(
        :'20150101' => { :energy_generated => 1239 },
        :'20150102' => { :energy_generated => 1523 }
      )

      expect(st).to have_been_requested
    end

    it 'adds a batch output output with loading delay' do
      body = 'data=20150101%2C1239%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C%3B20150102%2C1523%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C%2C'
      st = stub_request(:post, 'http://pvoutput.org/service/r2/addoutput.jsp').with(
        :body => body, :headers => headers
      ).to_return(:body => 'Load in progress', :status => 400).then.to_return(:status => 200)

      donation_client.add_batch_output(
        :'20150101' => { :energy_generated => 1239 },
        :'20150102' => { :energy_generated => 1523 }
      )

      expect(st).to have_been_requested.twice
    end
  end
end
# rubocop:enable Metrics/BlockLength,RSpec/FilePath


lib/pvoutput/client.rb
# frozen_string_literal: true

require 'httparty'

module PVOutput
  class Client
    include HTTParty
    base_uri 'pvoutput.org'
    debug_output $stdout if ENV['PVOUTPUT_DEBUG']

    def initialize(system_id, api_key, donation_mode = false)
      @system_id = system_id.to_s
      @api_key = api_key.to_s

      # The add_batch_output operation has a default limit of 1 sets of data in one
      # request, when you are using the donation mode the limit is 100 sets
      @batch_size = 1
      @batch_size = 100 if donation_mode

      self.class.headers 'X-Pvoutput-Apikey' => @api_key, 'X-Pvoutput-SystemId' => @system_id
    end

    # Helper method to post batch request to pvoutput that retries at the moment we get
    # a 400 error back with body containing 'Load in progress'
    def post_request(path, options = {}, &block)
      loop do
        response = self.class.post(path, options, &block)

        if response.code == 400 && response.body =~ /Load in progress/
          # We can't send data too fast, when the previous request is still loaded we
          # have to wait so sleep 10 seconds and try again
          sleep(10)
        elsif response.code == 200
          return
        else
          raise("Bad Post: #{response.body}")
        end
      end
    end

    # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
    def add_status(options)
      time = options[:when] || Time.now

      params = {
        'd' => time.strftime('%Y%m%d'),
        't' => time.strftime('%H:%M'),
      }

      params[:v1] = options[:energy_generated] if options[:energy_generated]
      params[:v2] = options[:power_generated] if options[:power_generated]
      params[:v3] = options[:energy_consumed] if options[:energy_consumed]
      params[:v4] = options[:power_consumed] if options[:power_consumed]
      params[:v5] = options[:temperature] if options[:temperature]
      params[:v6] = options[:voltage] if options[:voltage]
      params[:c1] = 1 if options[:cumulative] == true
      params[:n]  = 1 if options[:net] == true

      response = self.class.post('/service/r2/addstatus.jsp', :body => params)

      raise('Bad Post') unless response.code == 200
    end
    # rubocop:enable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity

    # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
    def add_output(options)
      params = {
      }

      params[:d] = options[:output_date]
      params[:g] = options[:energy_generated] if options[:energy_generated]
      params[:pp] = options[:peak_power] if options[:peak_power]
      params[:pt] = options[:peak_time] if options[:peak_time]
      params[:cd] = options[:condition] if options[:condition]
      params[:tm] = options[:min_temp] if options[:min_temp]
      params[:tx] = options[:max_temp] if options[:max_temp]
      params[:cm] = options[:comments] if options[:comments]
      params[:ip] = options[:import_peak] if options[:import_peak]
      params[:io] = options[:import_off_peak] if options[:import_off_peak]
      params[:is] = options[:import_shoulder] if options[:import_shoulder]
      params[:ih] = options[:import_high_shoulder] if options[:import_high_shoulder]
      params[:c] = options[:consumption] if options[:consumption]

      response = self.class.post('/service/r2/addoutput.jsp', :body => params)

      raise('Bad Post') unless response.code == 200
    end
    # rubocop:enable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity

    def add_batch_output(options)
      keys = %i[energy_generated energy_export energy_used]
      keys += %i[peak_power peak_time condition min_temp]
      keys += %i[max_temp comments import_peak import_off_peak]
      keys += %i[import_shoulder]

      options.to_a.each_slice(@batch_size) do |slice|
        data = ''
        slice.each do |entry|
          date, values = entry
          data += "#{date},#{keys.map { |key| values[key] }.join(',')};"
        end

        params = {
          :data => data.chop,
        }

        post_request('/service/r2/addoutput.jsp', :body => params)
      end
    end
  end
end
